src/main/webapp/app/exercises/programming/shared/instructions-render/extensions/programming-exercise-task.extension.ts

```diff
@@ -93,47 +113,46 @@ export class ProgrammingExerciseTaskExtensionWrapper implements ArtemisShowdownE
-                const taskRegex = /\[task\]\[.*\]\(.*\)({.*})?/g;
-                // E.g. Implement BubbleSort, testBubbleSort
-                const innerTaskRegex = /\[task\]\[(.*)\]\((.*)\)({(.*)})?/;
-                const tasks = text.match(taskRegex) || [];
-                const testsForTask: TaskArray = tasks
-                    .map((task) => {
-                        return task.match(innerTaskRegex);
-                    })
-                    .filter((testMatch) => !!testMatch && (testMatch.length === 3 || testMatch.length === 5))
-                    .map((testMatch: RegExpMatchArray) => {
-                        const nextIndex = this.taskIndex;
-                        this.taskIndex++;
-                        return {
-                            id: nextIndex,
-                            completeString: testMatch[0],
-                            taskName: testMatch[1],
-                            // split the names by "," only when there is not a closing bracket without a previous opening bracket
-                            tests: testMatch[2] ? testMatch[2].split(/,(?![^(]*?\))/).map((s) => s.trim()) : [],
-                        };
-                    });
-                const tasksWithParticipationId: TaskArrayWithExercise = {
-                    exerciseId: this.exercise.id!,
-                    tasks: testsForTask,
-                };
-                this.testsForTaskSubject.next(tasksWithParticipationId);
-                // Emit new found elements that need to be injected into html after it is rendered.
-                this.injectableElementsFoundSubject.next(() => {
-                    this.injectTasks(testsForTask);
-                });
-                return testsForTask.reduce(
-                    (acc: string, { completeString: task, id }): string =>
-                        // Insert anchor divs into the text so that injectable elements can be inserted into them.
-                        // Without class="d-flex" the injected components height would be 0.
-                        // Added zero-width space as content so the div actually consumes a line to prevent a <ol> display bug in Safari
-                        acc.replace(new RegExp(escapeStringForUseInRegex(task), 'g'), `<div class="pe-task-${id.toString()} d-flex">&#8203;</div>`),
-                    text,
-                );
+            filter: (problemStatement: string) => {
+                const tasks = Array.from(problemStatement.matchAll(this.taskRegex));
+                if (tasks) {
+                    return this.createTasks(problemStatement, tasks);
+                }
+                return problemStatement;
             },
         };
         return extension;
     }
+
+    private createTasks(problemStatement: string, tasks: RegExpMatchArray[]): string {
+        const testsForTask: TaskArray = tasks
+            .filter((testMatch) => testMatch?.length === 3)
```

Comment:
Where does the magic number 3 come from here? Maybe add a small comment (especially about when it will not be 3 - for some old test matches, I guess?)